### PR TITLE
Initial commit of cwe_checker plugin.

### PR DESCRIPTION
**Pull request by tbarabosch**

### CODE REVIEW

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            res[key] = {'plugin_version': plugin_version,
+                        'warnings': tmp_list}
+
+        return res
+
+    def _is_supported_arch(self, file_object):
+        arch_type = file_object.processed_analysis['file_type']['full'].lower()
+        for supported_arch in self.SUPPORTED_ARCHS:
+            if supported_arch in arch_type:
+                return True
```
Comment (jstucke) [inline src/plugins/analysis/cwe_checker/code/cwe_checker.py]: the builtins `any` and `all` are handy for cases like this, e.g.
```python
if any(supported_arch in arch_type for supported_arch in self.SUPPORTED_ARCHS):
    return True
```

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        file_type = file_object.processed_analysis['file_type']['full'].lower()
+        return 'elf' in file_type
+
+    def _do_full_analysis(self, file_object):
+        bap_command = self._build_bap_command(file_object)
+        output, return_code = execute_shell_command_get_return_code(
+            bap_command)
+        if return_code != 0:
+            logging.error('[%s] Could not communicate with Bap plugin: %i (%s).',
+                          self.NAME, return_code, output)
```
Comment (jstucke) [inline src/plugins/analysis/cwe_checker/code/cwe_checker.py]: see above

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        return file_object
+
+    def process_object(self, file_object):
+        '''
+        This function handles only ELF executable. Otherwise it returns an empty dictionary.
+        It calls the external BAP plugin cwe_checker.
+        '''
+        if not self._is_supported_file_type(file_object):
+            logging.debug('[%s] %s is not an ELF executable.',
+                          self.NAME, file_object.file_path)
```
Comment (jstucke) [inline src/plugins/analysis/cwe_checker/code/cwe_checker.py]: see above

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            file_object.processed_analysis[self.NAME] = {'summary': []}
+        elif not self._is_supported_arch(file_object):
+            logging.debug('[%s] %s\'s arch is not supported (%s)',
+                          self.NAME, file_object.file_path,
+                          file_object.processed_analysis['cpu_architecture']['summary'])
```
Comment (jstucke) [inline src/plugins/analysis/cwe_checker/code/cwe_checker.py]: see above

**Comment by jstucke** [comment]
Also, it would be nice if you could add some more tests because the coverage is relatively low: 
[coverage.zip](https://github.com/fkie-cad/FACT_core/files/2463503/coverage.zip)

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+            cwe_warning = splitted_line[1].replace(
+                'u32', '').replace(':', '')
+
+            cwe_name = self._remove_color(cwe_warning.split(')')[0]) + ')'
+            cwe_name = cwe_name.split('{')[0].strip() + ' ' + cwe_name.split('}')[1].strip()
+
+            plugin_version = cwe_warning.split('{')[1].split('}')[0]
+
+            cwe_message = ')'.join(cwe_warning.split(')')[1:])
+            cwe_message = cwe_message.replace('.', '').replace('32u', '')
```
[the remark this answers is not in the document]
Comment (tbarabosch) [reply src/plugins/analysis/cwe_checker/code/cwe_checker.py]: There will be JSON output soon. Therefore, I do not want to waste much time on parsing to be 100% perfect.

**Comment by tbarabosch** [comment]
@dorpvom There were still calls to the dockerized component. I monkeypatched them. There should be no calls to external programs. 
@jstucke I think I have tested the relevant components (parsers). The rest of this plugin is just wrapper code that calls cwe_checker (either locally or in Docker).

**Comment by tbarabosch** [comment]
I've pushed the required changes.

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        else:
+            bap_command = 'timeout --signal=SIGKILL {}m {} {} --pass=cwe-checker --cwe-checker-config={}/../internal/src/config.json'.format(
+                BAP_TIMEOUT,
+                PATH_TO_BAP,
+                file_object.file_path,
+                os.path.join(os.path.dirname(os.path.abspath(__file__))))
+        return bap_command
+
+    @staticmethod
+    def _parse_bap_output(output):
```
Comment (weidenba) [inline src/plugins/analysis/cwe_checker/code/cwe_checker.py]: This is a static method. It should be easy to test it?

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+            for cwe in values:
+                tmp_list.append(cwe.warning)
+                if not plugin_version:
+                    plugin_version = cwe.plugin_version
+            res[key] = {'plugin_version': plugin_version,
+                        'warnings': tmp_list}
+
+        return res
+
+    def _is_supported_arch(self, file_object):
```
Comment (weidenba) [inline src/plugins/analysis/cwe_checker/code/cwe_checker.py]: This should be easy to test as well.

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+            file_object.processed_analysis[self.NAME] = {'full': cwe_messages,
+                                                         'summary': list(cwe_messages.keys())}
+        return file_object
+
+    def process_object(self, file_object):
+        '''
+        This function handles only ELF executable. Otherwise it returns an empty dictionary.
+        It calls the external BAP plugin cwe_checker.
+        '''
+        if not self._is_supported_file_type(file_object):
```
Comment (weidenba) [inline src/plugins/analysis/cwe_checker/code/cwe_checker.py]: Please use the Whitlist functionality

File: src/plugins/analysis/cwe_checker/code/cwe_checker.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+    def _build_bap_command_for_modules_versions(self):
+        # unfortunately, there must be a dummy file passed to BAP, I chose /bin/true because it is damn small
+        if self.docker:
+            bap_command = 'docker run  cwe-checker:latest bap /bin/true --pass=cwe-checker --cwe-checker-module_versions=true'
+        else:
+            bap_command = '{} {} --pass=cwe-checker --cwe-checker-module_versions=true'.format(PATH_TO_BAP, '/bin/true')
+        return bap_command
+
+    def _build_bap_command(self, file_object):
```
Comment (weidenba) [inline src/plugins/analysis/cwe_checker/code/cwe_checker.py]: This should be easy to test?

File: src/plugins/analysis/cwe_checker/test/test_cwe_checker.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
+from test.unit.analysis.analysis_plugin_test_class import AnalysisPluginTest
+from ..code.cwe_checker import AnalysisPlugin, CweWarningParser
+
+
+class MockFileObject(object):
+
+    def __init__(self, file_path):
```
Comment (weidenba) [inline src/plugins/analysis/cwe_checker/test/test_cwe_checker.py]: This Mock is not used at all?

**Pull request merged**